Refactoring, contract dsl wasm compatibility, encode/decode fixes.

yglukhov opened the pull request:
I tried using `contract` dsl in a wasm web-app and it appeared impossible, as web3 is tied on much irrelevant stuff incompatible with wasm (like http client, monitoring, etc), so my initial goal was to separate `contract` macro to a separate module, to make it compatible with wasm. This involves some breaking changes in types (Sender and ContractCall) were affected in backward-incompatible manner. Instead of `Sender` we now have `ContractInstance[TSender]`, and `ContractCall[T]` was changed to `ContractInvocation[TSender, T]` respectively. Where `TSender` is just passed along from Instance to Invocation by contract dsl.

Also I've hit a bug in encoding/decodind dynamic objects, and while fixing it I thought it would be better to change the decoding/encoding to work with `seq[byte]` rather than `string` so there's that. Added  one more case to `EncodingTest`.

Also removed `Bool` type in favor of `bool`.

arnetheduck commented:
What would be backwards-compatibility consequences of this PR? Currently I can't compile nimbus with it which is probably trivially fixable, but in terms of actual behavior, are there any changes?

```
nimbus-eth2/vendor/nim-web3/web3/encoding.nim(40, 32) Error: expression has no address; maybe use 'unsafeAddr'
```

yglukhov commented:
> What would be backwards-compatibility consequences of this PR? 

Those are listed in the description types `Sender`, `ContractSender`, and `ContractCall` have changed. `Bool` was removed. Hopefully I'm not missing anything else.

> Currently I can't compile nimbus with it which is probably trivially fixable

Most likely the fix should be trivial. Unfortunately I'm pretty out of sync with nimbus build system to test this quickly. But I can look if you insist, or the fix appears to be not that trivial.

> but in terms of actual behavior, are there any changes?

There should not be any changes in high level behavior. Encoding/decoding should be more efficient, some bits moved to compile time, but that's it.

arnetheduck commented:
the changes look great in general, would love to get them merged

>  and while fixing it I thought it would be better to change the decoding/encoding to work with seq[byte] rather than string so there's that. 

this seems to be a breaking change - if anything, we need to retain both and maybe mark the `string` variations deprecated.. ie some degree of backwards compat is important so we can move back and forth between versions in case something breaks.

see https://github.com/status-im/nimbus-eth2/blob/9ec7b700a8fdcbc97b7f1a2f11c27be5cbe0db46/beacon_chain/el/el_manager.nim#L1489 for example - we read and use the string decoding.

```
nimbus-eth2/beacon_chain/el/el_manager.nim(1489, 21) Error: type mismatch: got <string, int, PubKeyBytes>
but expected one of:
func decode(input: openArray[byte]; baseOffset, offset: int;
            to: var DynamicBytes): int
  first type mismatch at position: 1
  required type for input: openArray[byte]
  but expression 'logData' is of type: string
```

yglukhov commented:
Ok, I believe I can make the pr somewhat backward compatible so that nimbus-eth2 will continue to work without any changes related to _this_ pr. However it seems that currently nimbus-eth2 doesn't link to the latest master version of nim-web3, and that causes me problems, specifically commit ebf4a12cbaf8d07c2a6c315adf6278a67c1fdc54. Would it be easy for you to update nimbus-eth2 to the latest master of nim-web3, so I can then fix backward compatibility on top of that?

Also, given that nim-web3 is already "backward incompatible", would it make sense to still preserve backward compatibility? Because like you said earlier, the changes required in nimbus-eth2 are rather trivial. I could as well make those, once you fix the question above.

arnetheduck commented:
> Also, given that nim-web3 is already "backward incompatible", would it make sense to still preserve backward compatibility? 

I think it makes sense to keep the decoders backwards-compatible while the specific web3 types might stay as-is - ie it looks to me like just keeping the string decoders should make things work decently (we can maybe deprecate them if they are .. unsound - I haven't considered this at depth)

4844 is still being developed (the spec is still changing), hence the "already incompatible" - ie that part of the object is WIP while _generally_ we strive to keep things as compatible as possible.

I'll get nimbus bumped.

arnetheduck commented:
>  nimbus-eth2 doesn't link to the latest master version of nim-web3

the `unstable` nimbus-eth2 branch has been updated already

yglukhov commented:
> the unstable nimbus-eth2 branch

That's the part I was missing :)

Ok, please see the latest commit. I've shuffled the types around just a bit to make the send/call apis backward compatible. I've added a string decoder for `DynamicBytes` to make nimbus-eth2 happy.

The string decoding api should be deprecated, because it is unsound. See that new `baseOffset` argument in byte decoding api? It is required to correctly decode nested dynamic types.

arnetheduck commented on web3/encoding.nim:
```diff
@@ -1,216 +1,178 @@
+func decode*(input: openarray[byte], baseOffset, offset: int, to: var StUint): int =
+  const meaningfulLen = to.bits div 8
+  let offset = offset + baseOffset
+  to = type(to).fromBytesBE(input[offset ..< offset + meaningfulLen])
```
hm, these decode functions assume sufficient input data, or they will panic - it's not a new issue, but looks risky

yglukhov replied:
Just keeping to status quo here.

arnetheduck commented on web3/encoding.nim:
```diff
@@ -1,216 +1,178 @@
 
-func decodeFixed(input: string, offset: int, to: var openArray[byte]): int =
-  let meaningfulLen = to.len * 2
+func copyArray(dst: var openarray[byte], src: openarray[byte]) {.inline.} =
```
`import stew/assign2`, `assign(dst, src)`

yglukhov replied:
Done

arnetheduck commented on web3/encoding.nim:
```diff
@@ -1,216 +1,178 @@
+func decode*(input: openarray[byte], baseOffset, offset: int, to: var StUint): int =
+  const meaningfulLen = to.bits div 8
+  let offset = offset + baseOffset
+  to = type(to).fromBytesBE(input[offset ..< offset + meaningfulLen])
```
```suggestion
  to = type(to).fromBytesBE(input.toOpenArray(offset, offset + meaningfulLen - 1))
```

`[]` causes an unnecessary seq allocation

arnetheduck commented on web3.nim:
```diff
@@ -153,6 +145,26 @@ proc subscribeForLogs*(w: Web3, options: JsonNode,
     result.historicalEventsProcessed = true
 
+proc addAddressAndSignatureToOptions(options: JsonNode, address: Address, topic: seq[byte]): JsonNode =
+  result = options
+  if result.isNil:
+    result = newJObject()
```
```suggestion
  result = if options.isNil: newJObject() else: options
```

arnetheduck commented:
something went wrong here, this OOMs:

```
#0  __GI___libc_write (fd=2, buf=0x13a6638, nbytes=14) at ../sysdeps/unix/sysv/linux/write.c:25
#1  0x00007ffff7d598f5 in _IO_new_file_write (f=0x7ffff7eac4e0 <_IO_2_1_stderr_>, data=0x13a6638, n=14) at fileops.c:1181
#2  0x00007ffff7d57a9f in new_do_write (fp=fp@entry=0x7ffff7eac4e0 <_IO_2_1_stderr_>, data=data@entry=0x13a6638 "out of memory\n", to_do=to_do@entry=14) at /usr/src/debug/glibc-2.38-10.fc39.x86_64/libio/libioP.h:1030
#3  0x00007ffff7d59af1 in _IO_new_file_xsputn (n=14, data=<optimized out>, f=0x7ffff7eac4e0 <_IO_2_1_stderr_>) at fileops.c:1255
#4  _IO_new_file_xsputn (f=0x7ffff7eac4e0 <_IO_2_1_stderr_>, data=<optimized out>, n=14) at fileops.c:1197
#5  0x00007ffff7d4d1e2 in __GI__IO_fwrite (buf=buf@entry=0x13a6638, size=size@entry=1, count=14, fp=fp@entry=0x7ffff7eac4e0 <_IO_2_1_stderr_>) at /usr/src/debug/glibc-2.38-10.fc39.x86_64/libio/libioP.h:1030
#6  0x00000000004838b8 in rawWrite (f=0x7ffff7eac4e0 <_IO_2_1_stderr_>, s=s@entry=0x13a6638 "out of memory\n") at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/memory.nim:47
#7  0x00000000004838e5 in raiseOutOfMem__system_4234 () at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/mmdisp.nim:48
#8  0x0000000000484e5a in osAllocPages__system_4259 (size=<optimized out>) at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/osalloc.nim:145
#9  getHugeChunk__system_5076 (size=4992398877043096379, a=0x7ffff7ccd778) at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/alloc.nim:620
#10 rawAlloc__system_5105 (requestedSize=4992398877043096347, a=0x7ffff7ccd778) at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/alloc.nim:752
#11 rawAlloc__system_5105 (a=0x7ffff7ccd778, requestedSize=4992398877043096347) at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/alloc.nim:694
#12 0x0000000001191dc2 in rawNewObj__system_6387.constprop.0 (typ=0x3dc81c0 <NTIseqLbyteT__6H5Oh5UUvVCLiakt9aTwtUQ_>, size=4992398877043096331, gch=<optimized out>) at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/gc.nim:458
#13 0x0000000000485d21 in newObj (typ=<optimized out>, size=4992398877043096331) at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/gc.nim:486
#14 0x0000000000488761 in newSeq (typ=typ@entry=0x3dc81c0 <NTIseqLbyteT__6H5Oh5UUvVCLiakt9aTwtUQ_>, len=len@entry=4992398877043096315) at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system/gc.nim:494
#15 0x000000000048daab in X5BX5D___OOZvendorZnim45ethZethZrlp_288 (s=0x7ffff5038780 "", sLen_0=576, x_0=...) at /nimbus-eth2/vendor/nimbus-build-system/vendor/Nim/lib/system.nim:2617
#16 0x0000000000b3afc3 in decode__OOZvendorZnim45web51Zweb51Zencoding_181 (offset=0, to=0x7fffffff6c78, baseOffset=32, inputLen_0=<optimized out>, input=0x7ffff5038520 "") at /nimbus-eth2/vendor/nim-web3/web3/encoding.nim:68
#17 decode__elZel95manager_14476 (offset=0, to=0x7fffffff6c48, baseOffset=32, inputLen_0=<optimized out>, input=0x7ffff5038520 "") at /nimbus-eth2/vendor/nim-web3/web3/encoding.nim:72
#18 decode__elZel95manager_14465 (to=0x7fffffff6c48, offset=64, input=0x7ffff5084500) at /nimbus-eth2/vendor/nim-web3/web3/encoding.nim:174
#19 depositEventsToBlocks__elZel95manager_14393 (depositsList=<optimized out>) at /nimbus-eth2/beacon_chain/el/el_manager.nim:1495
#20 syncBlockRange__elZel95manager_19869 (chronosInternalRetFuture=<optimized out>, ClE_0=<optimized out>) at /nimbus-eth2/beacon_chain/el/el_manager.nim:1956
#21 0x00000000004ef13b in futureContinue__OOZvendorZnim45chronosZchronosZinternalZasyncfutures_446 (fut=0x7ffff50473d0) at /nimbus-eth2/vendor/nim-chronos/chronos/internal/asyncfutures.nim:377
#22 0x00000000004e1426 in poll__OOZvendorZnim45chronosZchronosZinternalZasyncengine_8007 () at /nimbus-eth2/vendor/nim-chronos/chronos/internal/asyncengine.nim:150
#23 0x0000000001053ebd in run__nimbus95beacon95node_115364 (node=<optimized out>) at /nimbus-eth2/beacon_chain/nimbus_beacon_node.nim:1828
#24 start__nimbus95beacon95node_116612 (node=node@entry=0x7fffae1a1050) at /nimbus-eth2/beacon_chain/nimbus_beacon_node.nim:1881
#25 0x0000000001055fbe in doRunBeaconNode__nimbus95beacon95node_117694 (config=config@entry=0x7fffffff9980, rng_0=rng_0@entry=0x7ffff7c9bf50) at /nimbus-eth2/beacon_chain/nimbus_beacon_node.nim:2115
#26 0x0000000001057dfb in handleStartUpCmd__nimbus95beacon95node_178636 (config=config@entry=0x7fffffff9980) at /nimbus-eth2/beacon_chain/nimbus_beacon_node.nim:2214
#27 0x0000000001058cea in main__nimbus95beacon95node_178752 () at /nimbus-eth2/beacon_chain/nimbus_beacon_node.nim:2346
#28 0x00000000004204b1 in NimMain () at /nimbus-eth2/vendor/nim-libp2p/libp2p/protocols/rendezvous.nim:895
#29 main (argc=<optimized out>, args=<optimized out>, env=<optimized out>) at /nimbus-eth2/vendor/nim-libp2p/libp2p/protocols/rendezvous.nim:902
```

yglukhov commented:
@arnetheduck I've fixed the offending string decoder, and added a test to verify it yields exactly the same results as before. Anything else you see missing?
